keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1757 @@
+particular container. However, QoS-class resources are also different from
+those other resources because they are used to assign a _class identifier_,
+rather than to declare a specific amount of capacity that is allocated.
+
+Main characteristics of the new resource type (and the technologies they are
+aimed at enabling) are:
+
+- multiple containers can be assigned to the same class of a certain type of
+  QoS-class resource
+- QoS-class resources are represented by an enumerable set of class identifiers
+- each type of QoS-class resource has an independent set of class identifiers
+
+With QoS-class resources, Pods and their containers can request opaque
+QoS-class identifiers (classes) of certain QoS mechanism (QoS-class resource
+type). Kubelet relays this information to the container runtime, without
+directing how the request is enforced in the underlying system.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+This enhancement proposal aims at improving the quality of service of
+applications by making available a new type of resource control mechanism in
+Kubernetes. Certain types of resources are inherently shared by applications,
+e.g. cache, memory bandwidth and disk I/O. While there are technologies for
+controlling how these resources are shared between applications, there is
+currently no meaningful way to support these technologies in Kubernetes. This
+proposal suggests to address the issue above in a generalized way by extending
+the Kubernetes resource model with a new type of resources, i.e. QoS-class
+resources.
+
+This KEP identifies two technologies that can immediately be enabled with
+QoS-class resources. However, these are just two examples  and the proposed
+changes are generic (and not tied to these two QoS-class resource types in any
+way), making it easier to implement new QoS-class resource types.
+
+[Intel RDT][intel-rdt] implements a class-based mechanism for controlling the
+cache and memory bandwidth QoS of applications. All processes in the same
+hardware class share a portion of cache lines and memory bandwidth. RDT
+provides a way for mitigating noisy neighbors and fulfilling SLAs. In Linux
+control happens via resctrl -- a pseudo-filesystem provided by the kernel which
+makes it virtually agnostic of the hardware architecture. The OCI runtime-spec
+has supported Intel RDT for a while already. Other hardware vendors have
+comparable technologies which use the same [resctrl interface][linux-resctrl].
+
+The Linux Block IO controller parameters depend very heavily on the underlying
```

Comment:
This is probably the best example of the configuration: https://github.com/intel/goresctrl/blob/main/doc/blockio.md

It can be used for anything seen as block device storage by the kernel. Basically local disks. The configuration supports globbing (for device paths) but is sort of fixed as it's specified in a configuration file

Reply:
@marquiz looks interesting, I was also looking for same sort of solution to support the QOS for block devices, We do have cephcsi driver which maps the rbd image as a block device, With this KEP is it possible to set the QOS for the device created as part of csi driver and mount it just a filesystem/block PVC to the application pod?  As the devices get created dynamically by the csi driver (with random device numbers) when mounting the PVC how admin can pre-create the class and set QoS for these devices? sorry if my question is confusing

Reply:
@Madhu-1 yes, that should work. However, there is a caveat that if you have also some other CSI providers exposing the volumes as rbd devices, those would also get throttled (in case you only wanted to throttle the cephcsi volumes). 

As a note: I think the eventual end goal would be to implement the blockio as an [NRI](https://github.com/containerd/nri) plugin. The plugin would get all relevant information (about volumes etc) at container creation time and set cgroup parameters for the exact devices that we want without any globbing.

Reply:
@marquiz got it, if we have any application pod that doesn't have any class name set but is using the RBD PVC , Will those pods will also get affected by this as the device mapped will come in the range specified above or limits get applied to the pod which is having only the class set (based on the class name) in the yaml.

Reply:
> if we have any application pod that doesn't have any class name set but is using the RBD PVC , Will those pods will also get affected by this as the device mapped will come in the range specified above or limits get applied to the pod which is having only the class set (based on the class name) in the yaml.

@Madhu-1 no, only those pods that have requested the class will be affected by this class.

Reply:
got it, Thank you @marquiz for the confirmation.

Reply:
@marquiz one last questions 🗡️ will this work on both cgroup v1 and cgroup v2? or it works only with cgroup v2

Reply:
@marquiz in case if you have missed the above question.